src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.MultiSegment.cs

```diff
@@ -0,0 +1,2408 @@
+                }
+                else if (ConsumeValueMultiSegment(first))
+                {
+                    goto Done;
+                }
+
+                return false;
+
+            Done:
+                // Cannot use HasMoreData since the JSON payload contains a single, non-primitive value
+                // and hence must be handled differently.
+                if (_consumed >= (uint)_buffer.Length)
+                {
+                    return true;
+                }
+
+                if (_buffer[_consumed] <= JsonConstants.Space)
+                {
+                    SkipWhiteSpaceMultiSegment();
+                    if (_consumed >= (uint)_buffer.Length)
+                    {
+                        return true;
+                    }
+                }
+
+                if (_readerOptions.CommentHandling != JsonCommentHandling.Disallow)
+                {
+                    if (_readerOptions.CommentHandling == JsonCommentHandling.Allow)
+                    {
+                        // This is necessary to avoid throwing when the user has 1 or more comments as the first token
+                        // OR if there is a comment after a single, non-primitive value.
+                        // In this mode, ConsumeValue consumes the comment and we need to return it as a token.
+                        // along with future comments in subsequeunt reads.
+                        if (_tokenType == JsonTokenType.Comment || _buffer[_consumed] == JsonConstants.Slash)
+                        {
+                            return true;
+                        }
+                    }
+                    else
+                    {
+                        Debug.Assert(_readerOptions.CommentHandling == JsonCommentHandling.Skip);
+                        if (_tokenType == JsonTokenType.StartObject || _tokenType == JsonTokenType.StartArray)
+                        {
+                            _isNotPrimitive = true;
+                        }
+                        return true;
+                    }
+                }
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedEndAfterSingleJson, _buffer[_consumed]);
+            }
+            return true;
+        }
+
+        private void SkipWhiteSpaceMultiSegment()
+        {
+            while (true)
+            {
+                // Create local copy to avoid bounds checks.
+                ReadOnlySpan<byte> localBuffer = _buffer;
+                for (; _consumed < localBuffer.Length; _consumed++)
+                {
+                    byte val = localBuffer[_consumed];
+
+                    // JSON RFC 8259 section 2 says only these 4 characters count, not all of the Unicode defintions of whitespace.
+                    if (val != JsonConstants.Space &&
+                        val != JsonConstants.CarriageReturn &&
+                        val != JsonConstants.LineFeed &&
+                        val != JsonConstants.Tab)
```

Comment:
Rather than repeat all of this, would

```C#
private void SkipWhiteSpaceMultiSegment()
{
    while (true)
    {
        SkipWhitespace();

        if (_consumed < _buffer.Length)
        {
            break;
        }

        if (!GetNextSpan())
        {
            break;
        }
    }
}
```

work?